include/pybind11/factory_support.h

```diff
@@ -0,0 +1,280 @@
+    // cpp pointer is actually a polymorphic alias instance.  If it isn't, but we can construct the
+    // alias via an `Alias(Cpp &&)` constructor, we do.  Otherwise we throw a type error.
+    template <typename Class, enable_if_t<Class::has_alias, int> = 0>
+    static void construct(value_and_holder &v_h, Cpp<Class> *ptr, PyTypeObject *cl_type) {
+        no_nullptr(ptr);
+        if (Py_TYPE(v_h.inst) != cl_type) {
+            // Inherited from on the Python side: an alias instance is required
+            if (!is_alias<Class>(ptr)) {
+                // We need to do destruction through the holder because it might have a special
+                // deleter, or might be shared_from_this.
+                deallocate(v_h) = ptr;
+                v_h.set_instance_registered(true); // To prevent init_instance from registering it
+                v_h.type->init_instance(v_h.inst, nullptr); // Set up the holder
+                auto temp_holder = std::move(v_h.holder<Holder<Class>>()); // Steal the holder
+                v_h.type->dealloc(v_h); // Destroys the moved-out holder remains, resets value ptr to null
+                v_h.set_instance_registered(false);
+
+                if (construct_alias_from_cpp<Class>(v_h, std::move(*ptr)))
+                    return;
+                else
+                    throw type_error("pybind11::init(): factory function pointer could not be cast or "
+                                     "converted to an alias instance");
+            }
+        }
+        deallocate(v_h) = ptr;
+    }
+
+    // Holder return: copy its pointer, and move or copy the returned holder into the new instance's
+    // holder.  This also handles types like std::shared_ptr<T> and std::unique_ptr<T> where T is a
+    // derived type (through those holder's implicit conversion from derived class holder constructors).
+    template <typename Class>
+    static void construct(value_and_holder &v_h, Holder<Class> holder, PyTypeObject *cl_type) {
+        auto *ptr = holder_helper<Holder<Class>>::get(holder);
+        // If we need an alias, check that the held pointer is actually an alias instance
+        if (Class::has_alias && Py_TYPE(v_h.inst) != cl_type && !is_alias<Class>(ptr))
+            throw type_error("pybind11::init(): construction failed: returned holder-wrapped instance "
+                             "is not an alias instance");
+
+        deallocate(v_h) = ptr;
+        v_h.type->init_instance(v_h.inst, &holder);
+    }
+
+    // return-by-value version 1: returning a cpp class by value when there is no alias
+    template <typename Class, enable_if_t<!Class::has_alias, int> = 0>
+    static void construct(value_and_holder &v_h, Cpp<Class> &&result, PyTypeObject *) {
+        static_assert(std::is_move_constructible<Cpp<Class>>::value,
+            "pybind11::init() return-by-value factory function requires a movable class");
+        new (preallocate(v_h)) Cpp<Class>(std::move(result));
```

Comment:
`preallocate` doesn't do anything if the value is already preallocated, so if you mix old-style placement-new `__init__`s with factory inits you get fewer allocation/deallocations.  If placement-new constructors are rare/deprecated, it seems a reasonable approach though.

(And on a side note, non-factory inits have a pretty big issue with the constructor being invoked multiple time if being used under multiple inheritance, so discouraging them is probably a good thing).